File: app/src/main/java/com/checkin/app/checkin/Data/BaseViewModel.java
package com.checkin.app.checkin.Data;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.checkin.app.checkin.Utility.ProblemHandler;
import com.checkin.app.checkin.Utility.SourceMappedLiveData;
import com.fasterxml.jackson.databind.node.ObjectNode;

public abstract class BaseViewModel extends AndroidViewModel {
    private Application mApplication;
    protected SourceMappedLiveData<Resource<ObjectNode>> mData = createNetworkLiveData();
    ProblemHandler problemHandler;

    public BaseViewModel(@NonNull Application application) {
        super(application);
        mApplication = application;
    }

    public LiveData<Resource<ObjectNode>> getObservableData() {
        return mData;
    }

    protected <T> SourceMappedLiveData<Resource<T>> createNetworkLiveData() {
        return new SourceMappedLiveData<>(resource -> {
            problemHandler = new ProblemHandler(mApplication);
            if (ProblemHandler.handleProblems(mApplication, resource))
                return null;
            return resource;
        });
    }

    public void resetObservableData() {
        mData.setValue(null);
    }

    public abstract void updateResults();
}

File: app/src/main/java/com/checkin/app/checkin/data/BaseViewModel.kt
package com.checkin.app.checkin.data

import android.app.Application
import androidx.annotation.CallSuper
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import com.checkin.app.checkin.data.network.ProblemHandler
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.utility.SourceMappedLiveData
import com.fasterxml.jackson.databind.node.ObjectNode

interface IBaseViewModel {
    fun <T> createNetworkLiveData(): SourceMappedLiveData<Resource<T>>
}

abstract class BaseViewModel(private val mApplication: Application) : AndroidViewModel(mApplication), IBaseViewModel {
    protected var mData = createNetworkLiveData<ObjectNode>()

    val observableData: LiveData<Resource<ObjectNode>>
        get() = mData

    final override fun <T> createNetworkLiveData(): SourceMappedLiveData<Resource<T>> = SourceMappedLiveData { resource ->
        if (ProblemHandler.handleProblems(mApplication, resource)) null else resource
    }

    fun resetObservableData() {
        mData.value = null
    }

    abstract fun updateResults()

    open fun fetchMissing() {}

    @CallSuper
    open fun resetData() {
    }
}
